src/main/java/com/example/demo/services/BookingService.java

```diff
@@ -0,0 +1,272 @@
+
+    public List<MovieModel> moviesInTimeInterval(Date start, Date end) {
+
+        List<MovieModel> list = movieMapper.toModel(
+                        movieRepository.findByStartBeforeAndEndAfterOrderByStart(end, start))
+                .stream()
+                .map(Optional<MovieModel>::get)
+                .collect(Collectors.toList());
+
+        return list;
+    }
+
+
+    public Optional<MovieDetails> movieDetails(Long id) {
+        MovieModel movie = movieMapper.toModel(movieRepository.findById(id).orElseThrow()).orElseThrow();
+
+        List<SeatModel> allSeats = movie.getRoom().getSeats();
+
+        List<SeatModel> availableSeats = allSeats.stream()
+                .filter(item -> !movie.getReservations().contains(item.getId()))
+                .collect(Collectors.toList());
+
+        return Optional.of(new MovieDetails(movie, availableSeats));
+    }
+
+
+    private List<SeatEntity> deepCopySeatList(List<SeatEntity> s) {
+        ArrayList<SeatEntity> result = new ArrayList<>();
+        for (SeatEntity seat : s) {
+            result.add(new SeatEntity(seat));
+        }
+        return result;
+    }
+
+    public Optional<ReservationResponseModel> makeReservation(ReservationRequestModel request) {
+
+        if (!request.validIdTicketTypes()) {
+            throw new IllegalArgumentException("Invalid ticket types");
+        }
+
+        List<Long> sentSeatIds = request.getSeats().stream().map(Pair::getFirst).collect(Collectors.toList());
+        if (sentSeatIds.size() == 0) {
+            throw new IllegalArgumentException("0 seats selected");
+        }
+
+        MovieEntity movie = movieRepository.findById(request.getMovieId()).orElseThrow();
+
+        List<Long> reservedSeatIds = movie.getReservations().stream()
+                .map(ReservationEntity::getSeatId).collect(Collectors.toList());
+
+        List<SeatEntity> availableSeats = movie.getRoom().getSeats().stream()
+                .filter(item -> !reservedSeatIds.contains(item.getId()))
+                .collect(Collectors.toList());
+
+        List<Long> availableSeatIds = availableSeats.stream().map(SeatEntity::getId).collect(Collectors.toList());
+
+        // check if desired seats are empty
+        for (Pair<Long, String> seat : request.getSeats()) {
+            if (!availableSeatIds.contains(seat.getFirst())) {
+                throw new IllegalArgumentException("Seat is already taken");
+            }
+        }
+
+        List<Long> seatsFromReservation =
+                request.getSeats().stream()
+                        .map(Pair<Long, String>::getFirst)
+                        .collect(Collectors.toList());
+
+        // free after reservation
+        availableSeatIds.removeAll(seatsFromReservation);
+
+
+        //check if there exists a singular free seat between reserved seats
+
+        List<SeatEntity> sortedAllSeats = deepCopySeatList(movie.getRoom().getSeats());
+        Collections.sort(sortedAllSeats, new SortSeatEntity());
```

Comment:
drugi parametr to entity?

Reply:
nie to klasa, która implementuje funkcję compare dla typu SeatEntity, przydałaby się lepsza nazwa